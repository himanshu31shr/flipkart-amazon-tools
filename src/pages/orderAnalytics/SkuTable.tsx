import React from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Collapse,
} from "@mui/material";

type SkuTable = {
  skus: { [sku: string]: { orders: number; returns: number, title: string } };
  open: boolean;
};

export const SkuTable: React.FC<SkuTable> = ({ skus, open }) => (
  <TableRow>
    <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={7}>
      <Collapse in={open} timeout="auto" unmountOnExit>
        <Table size="small">
          <TableHead>
            <TableRow>
              <TableCell>SKU</TableCell>
              <TableCell>Title</TableCell>
              <TableCell>Orders</TableCell>
              <TableCell>Returns</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {Object.entries(skus).map(([sku, counts]) => (
              <TableRow key={sku}>
                <TableCell>{sku}</TableCell>
                <TableCell>{counts.title}</TableCell>
                <TableCell>{counts.orders}</TableCell>
                <TableCell>{counts.returns}</TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </Collapse>
    </TableCell>
  </TableRow>
);
