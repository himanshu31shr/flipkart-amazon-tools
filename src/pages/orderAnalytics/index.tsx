import React, { useEffect, useState } from "react";
import {
  Box,
  Container,
  Grid,
  Paper,
  Typography,
  CircularProgress,
  IconButton,
  Badge,
  Button,
} from "@mui/material";
import FilterListIcon from "@mui/icons-material/FilterList";
import CalendarTodayIcon from "@mui/icons-material/CalendarToday";
import { format } from "date-fns";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../../store";
import { fetchAllOrdersForAnalytics } from "../../store/slices/allOrdersForAnalyticsSlice";
import { fetchProducts } from "../../store/slices/productsSlice";
import { fetchCategories } from "../../store/slices/categoriesSlice";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { AdapterDateFns } from "@mui/x-date-pickers/AdapterDateFns";
import CategoryDistributionTable from "./components/CategoryDistributionTable";
import CategoryOrdersChart from "./components/CategoryOrdersChart";
import OrderMetrics from "./components/OrderMetrics";
import SkuOrdersChart from "./components/SkuOrdersChart";
import TopProductsTable from "./components/TopProductsTable";
import DateRangeFilter from "./components/DateRangeFilter";
import FilterPopover from "./components/FilterPopover";
import { useOrderFilters } from "./hooks/useOrderFilters";

const OrderAnalytics: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const {
    items: allOrders,
    loading: allOrdersLoading,
    error: allOrdersError,
  } = useSelector((state: RootState) => state.allOrdersForAnalytics);
  const {
    items: products,
    loading: productsLoading,
    error: productsError,
  } = useSelector((state: RootState) => state.products);
  const {
    items: categories,
    loading: categoriesLoading,
    error: categoriesError,
  } = useSelector((state: RootState) => state.categories);

  // State for popovers
  const [filterAnchorEl, setFilterAnchorEl] = useState<null | HTMLElement>(
    null
  );
  const [dateAnchorEl, setDateAnchorEl] = useState<null | HTMLElement>(null);

  // Use custom hook for filters
  const {
    filterState,
    skuOptions,
    platformOptions,
    productOptions,
    filteredOrders,
    updateFilter,
    clearFilters,
  } = useOrderFilters({ allOrders, categories, products });

  useEffect(() => {
    if (allOrders.length === 0 && !allOrdersLoading && !allOrdersError) {
      dispatch(fetchAllOrdersForAnalytics());
    }
    if (products.length === 0 && !productsLoading && !productsError) {
      dispatch(fetchProducts({}));
    }
    if (categories.length === 0 && !categoriesLoading && !categoriesError) {
      dispatch(fetchCategories());
    }
  }, []);

  const isLoading = allOrdersLoading || productsLoading || categoriesLoading;
  const error = allOrdersError || productsError || categoriesError;

  // Popover handlers
  const handleFilterClick = (event: React.MouseEvent<HTMLElement>) => {
    setFilterAnchorEl(event.currentTarget);
  };
  const handleFilterClose = () => {
    setFilterAnchorEl(null);
  };
  const handleDateClick = (event: React.MouseEvent<HTMLElement>) => {
    setDateAnchorEl(event.currentTarget);
  };
  const handleDateClose = () => {
    setDateAnchorEl(null);
  };

  // Active filter count
  const activeFilterCount = [
    filterState.selectedCategory,
    filterState.selectedSku,
    filterState.selectedPlatform,
    filterState.selectedProduct,
  ].filter(Boolean).length;

  // Date range summary
  const dateSummary = `${format(
    filterState.dateRange.startDate,
    "dd MMM yyyy"
  )} - ${format(filterState.dateRange.endDate, "dd MMM yyyy")}`;

  if (error) {
    return (
      <Container maxWidth="xl" sx={{ mt: 4, mb: 4 }}>
        <Typography color="error" variant="h6">
          Error: {error}
        </Typography>
      </Container>
    );
  }

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <Container maxWidth="xl" sx={{ mt: 4, mb: 4 }}>
        <Box
          display="flex"
          justifyContent="space-between"
          alignItems="center"
          mb={3}
        >
          <Typography variant="h4">Order Analytics</Typography>
          <Box display="flex" gap={2}>
            <Button
              variant="outlined"
              startIcon={<CalendarTodayIcon />}
              onClick={handleDateClick}
              sx={{ textTransform: "none" }}
            >
              {dateSummary}
            </Button>
            <IconButton
              onClick={handleFilterClick}
              color={activeFilterCount > 0 ? "primary" : "default"}
              sx={{ ml: 1 }}
            >
              <Badge badgeContent={activeFilterCount} color="primary">
                <FilterListIcon />
              </Badge>
            </IconButton>
          </Box>
        </Box>

        <DateRangeFilter
          dateRange={filterState.dateRange}
          onDateRangeChange={(startDate, endDate) =>
            updateFilter("dateRange", { startDate, endDate })
          }
          anchorEl={dateAnchorEl}
          onClose={handleDateClose}
        />

        <FilterPopover
          anchorEl={filterAnchorEl}
          onClose={handleFilterClose}
          categories={categories}
          selectedCategory={filterState.selectedCategory}
          selectedSku={filterState.selectedSku}
          selectedPlatform={filterState.selectedPlatform}
          selectedProduct={filterState.selectedProduct}
          skuOptions={skuOptions}
          platformOptions={platformOptions}
          productOptions={productOptions}
          onCategoryChange={(category) =>
            updateFilter("selectedCategory", category)
          }
          onSkuChange={(sku) => updateFilter("selectedSku", sku)}
          onPlatformChange={(platform) =>
            updateFilter("selectedPlatform", platform)
          }
          onProductChange={(product) =>
            updateFilter("selectedProduct", product)
          }
          onClearFilters={clearFilters}
        />

        {isLoading ? (
          <Box
            display="flex"
            justifyContent="center"
            alignItems="center"
            minHeight="400px"
          >
            <CircularProgress />
          </Box>
        ) : (
          <>
            <OrderMetrics
              orders={filteredOrders}
              products={products}
              categories={categories}
            />
            <Paper sx={{ p: 2 }}>
              <Grid container spacing={3}>
                <Grid item xs={12} md={7}>
                  <Typography variant="h6" gutterBottom>
                    Category Distribution Details
                  </Typography>
                  <CategoryDistributionTable
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
                <Grid item xs={12} md={5}>
                  <Typography variant="h6" gutterBottom>
                    Orders by Category
                  </Typography>
                  <CategoryOrdersChart
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
              </Grid>
            </Paper>
            <Paper sx={{ p: 2, mt:4, height: "100%" }}>
              <Grid container spacing={3}>
                <Grid item xs={12} md={7}>
                  <Typography variant="h6" gutterBottom>
                    Top Products
                  </Typography>
                  <TopProductsTable
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
                <Grid item xs={12} md={5}>
                  <Typography variant="h6" gutterBottom>
                    Orders by SKU
                  </Typography>
                  <SkuOrdersChart
                    orders={filteredOrders}
                    categories={categories}
                  />
                </Grid>
              </Grid>
            </Paper>
          </>
        )}
      </Container>
    </LocalizationProvider>
  );
};

export default OrderAnalytics;

/* TODO: Create the CategoryDistributionChart component */
