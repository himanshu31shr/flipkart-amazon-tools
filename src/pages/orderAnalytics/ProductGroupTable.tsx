import React, { useState } from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
  IconButton,
  TableSortLabel,
  TableFooter,
  Typography,
} from "@mui/material";
import { KeyboardArrowDown, KeyboardArrowUp } from "@mui/icons-material";
import { SkuTable } from "./SkuTable";
import { Group } from "./types/group.type";

type ProductGroupTableProps = {
  groupedData: Group[];
  openRow: string | null;
  handleRowClick: (group: string) => void;
};

type Order = "asc" | "desc";

export const ProductGroupTable: React.FC<ProductGroupTableProps> = ({
  groupedData,
  openRow,
  handleRowClick,
}) => {
  const [order, setOrder] = useState<Order>("asc");
  const [orderBy, setOrderBy] = useState<keyof Group>("group");

  const handleRequestSort = (property: keyof Group) => {
    const isAsc = orderBy === property && order === "asc";
    setOrder(isAsc ? "desc" : "asc");
    setOrderBy(property);
  };

  const sortedData = [...groupedData].sort((a, b) => {
    if (orderBy === "group") {
      return order === "asc"
        ? a.group.localeCompare(b.group)
        : b.group.localeCompare(a.group);
    } else {
      return order === "asc"
        ? Number(a[orderBy]) - Number(b[orderBy])
        : Number(b[orderBy]) - Number(a[orderBy]);
    }
  });

  return (
    <TableContainer component={Paper}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell />
            <TableCell>
              <TableSortLabel
                active={orderBy === "group"}
                direction={orderBy === "group" ? order : "asc"}
                onClick={() => handleRequestSort("group")}
              >
                Group
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "orders"}
                direction={orderBy === "orders" ? order : "asc"}
                onClick={() => handleRequestSort("orders")}
              >
                Orders
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active={orderBy === "returns"}
                direction={orderBy === "returns" ? order : "asc"}
                onClick={() => handleRequestSort("returns")}
              >
                Returns
              </TableSortLabel>
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {sortedData.map((group) => (
            <React.Fragment key={group.group}>
              <TableRow onClick={() => handleRowClick(group.group)}>
                <TableCell>
                  <IconButton>
                    {openRow === group.group ? (
                      <KeyboardArrowUp />
                    ) : (
                      <KeyboardArrowDown />
                    )}
                  </IconButton>
                </TableCell>
                <TableCell>
                  <strong>{group.group.replaceAll("_", " ")}</strong>
                </TableCell>
                <TableCell>{group.orders}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
                <TableCell>{group.returns}</TableCell>
              </TableRow>
              <SkuTable skus={group.skus} open={openRow === group.group} />
            </React.Fragment>
          ))}
        </TableBody>
        <TableFooter>
          <TableRow>
            <TableCell colSpan={2}>
              <Typography variant="h6" gutterBottom component="div">
                Totals
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h6" gutterBottom component="div">
                {sortedData.reduce((acc, curr) => acc + curr.orders, 0)}
              </Typography>
            </TableCell>
            <TableCell>
              <Typography variant="h6" gutterBottom component="div">
                {sortedData.reduce((acc, curr) => acc + curr.returns, 0)}
              </Typography>
            </TableCell>
          </TableRow>
        </TableFooter>
      </Table>
    </TableContainer>
  );
};
