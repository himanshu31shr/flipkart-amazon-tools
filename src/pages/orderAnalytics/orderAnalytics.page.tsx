import CloseIcon from "@mui/icons-material/Delete";
import { Alert, Box, CircularProgress, IconButton } from "@mui/material";
import Grid from "@mui/material/Grid2";
import Papa from "papaparse";
import React, { useEffect, useState } from "react";
import { deleteDatabase, getOrders } from "../../storage/db";
import { FileInput } from "../home/file-input";
import { ProductGroupTable } from "./ProductGroupTable";
import { OrderAggregationService } from "./services/orderAggregation.service";
import { Group } from "./types/group.type";
import { OrderItem } from "./types/order.type";

export const OrderAnalyticsPage = () => {
  const [orderItems, setOrderItems] = useState<OrderItem[]>([]);
  const [groupedData, setGroupedData] = useState<Group[]>([]);
  const [timeline, setTimeLine] = useState<
    Partial<{ start: string; end: string }>
  >({});

  const [fetchingDb, setFetchingDb] = useState(false);

  const fetchOrders = async () => {
    setFetchingDb(true);
    const savedOrders = await getOrders();
    if (savedOrders.length > 0) {
      setOrderItems(savedOrders);
      const service = new OrderAggregationService(savedOrders as OrderItem[]);
      setGroupedData(await service.process());
      setTimeLine(service.getDates());
    }
    setFetchingDb(false);
  };

  const deleteDb = () => {
    setFetchingDb(true);
    deleteDatabase();
    setOrderItems([]);
    setGroupedData([]);
    setTimeLine({});
    setFetchingDb(false);
  };

  useEffect(() => {
    fetchOrders();
  }, []);

  const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFetchingDb(true);
    const file = event.target.files?.[0];
    if (file) {
      Papa.parse(file, {
        complete: async (results: Papa.ParseResult<OrderItem>) => {
          const data = results.data as OrderItem[];
          setOrderItems(data);
          const service = new OrderAggregationService(data);
          setGroupedData(await service.process());
          setTimeLine(service.getDates());
          setFetchingDb(false);
        },
        header: true,
        transform: (value: string, field: string) => {
          if (field === "sku") {
            value = value.replaceAll(/["]|sku:/gi, "");
          }

          return value;
        },
        skipEmptyLines: true,
      });
    }
  };

  return (
    <Loader showLoader={fetchingDb}>
      <Grid
        container
        spacing={2}
        justifyContent={"space-between"}
        marginTop={2}
        marginBottom={2}
      >
        <Box>
          <FileInput
            onChange={handleFileUpload}
            name="Flipkart Report"
            accepts=".csv"
            selected={orderItems == null}
          />
        </Box>
        <Box>
          <IconButton onClick={() => deleteDb()} data-testid="delete-data">
            <CloseIcon />{" "}
          </IconButton>
        </Box>
        <Grid size={12}>
          {timeline.start && timeline.end && (
            <Alert color="info">
              Showing data between {new Date(timeline.start).toDateString()} &{" "}
              {new Date(timeline.end).toDateString()}
            </Alert>
          )}
        </Grid>
      </Grid>
      {groupedData.length > 0 && (
        <ProductGroupTable
          groupedData={groupedData}
        />
      )}
    </Loader>
  );
};

const Loader = ({
  showLoader = false,
  children,
}: {
  showLoader: boolean;
  children: React.ReactNode;
}) => {
  return (
    <>
      {showLoader && (
        <Grid spacing={10} marginTop={2} alignItems={"center"}>
          <CircularProgress size={16} />{" "}Building data...
        </Grid>
      )}
      {children}
    </>
  );
};
