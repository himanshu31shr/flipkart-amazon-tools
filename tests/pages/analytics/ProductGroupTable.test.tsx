import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { ProductGroupTable } from "../../../src/pages/orderAnalytics/ProductGroupTable";
import { Group } from "../../../src/pages/orderAnalytics/types/group.type";

const mockGroupedData: Group[] = [
  {
    group: "Group A",
    orders: 10,
    returns: 2,
    skus: {
      sku1: {
        orders: 5,
        returns: 1,
        title: "",
      },
      sku2: {
        orders: 5,
        returns: 1,
        title: "",
      },
    },
  },
  {
    group: "Group B",
    orders: 20,
    returns: 4,
    skus: {
      sku3: {
        orders: 10,
        returns: 2,
        title: "",
      },
      sku4: {
        orders: 10,
        returns: 2,
        title: "",
      },
    },
  },
];

describe("ProductGroupTable", () => {
  test("renders without crashing", () => {
    expect(
      render(<ProductGroupTable groupedData={mockGroupedData} />).container
    ).toMatchSnapshot();
  });

  test("renders table headers", () => {
    render(<ProductGroupTable groupedData={mockGroupedData} />);
    const groupHeader = screen.getByTestId("group-header");
    const ordersHeader = screen.getByTestId("order-header");
    const returnsHeader = screen.getByTestId("return-header");
    expect(groupHeader).toBeInTheDocument();
    expect(ordersHeader).toBeInTheDocument();
    expect(returnsHeader).toBeInTheDocument();
  });

  test("sorts data correctly by group", () => {
    render(<ProductGroupTable groupedData={mockGroupedData} />);
    const groupHeader = screen.getByTestId("group-header");
    fireEvent.click(groupHeader);
    const firstRow = screen.getAllByRole("row")[1];
    expect(firstRow.children.item(1)).toHaveTextContent("Group B");
    fireEvent.click(groupHeader);
    const secondRow = screen.getAllByRole("row")[1];
    expect(secondRow.children.item(1)).toHaveTextContent("Group A");
  });

  test("sorts data correctly by orders", () => {
    render(<ProductGroupTable groupedData={mockGroupedData} />);
    const ordersHeader = screen.getByTestId("order-header");
    fireEvent.click(ordersHeader);
    const firstRow = screen.getAllByRole("row")[1];
    expect(firstRow).toHaveTextContent("Group A");
    fireEvent.click(ordersHeader);
    const secondRow = screen.getAllByRole("row")[1];
    expect(secondRow).toHaveTextContent("Group B");
  });

  test("expands and collapses row on click", () => {
    render(<ProductGroupTable groupedData={mockGroupedData} />);
    const firstRow = screen.getAllByRole("row")[1];
    fireEvent.click(firstRow);
    const expandedRow = screen.getByText(/sku1/i);
    expect(expandedRow).toBeInTheDocument();
  });
});
