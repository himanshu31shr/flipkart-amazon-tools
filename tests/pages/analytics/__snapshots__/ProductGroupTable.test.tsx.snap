// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductGroupTable renders without crashing 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-y2ff7i-MuiPaper-root-MuiTableContainer-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <table
      class="MuiTable-root css-1xwxv7r-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1xtozoh-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          />
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root Mui-active MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              data-testid="group-header"
              role="button"
              tabindex="0"
            >
              Group
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              data-testid="order-header"
              role="button"
              tabindex="0"
            >
              Orders
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              data-testid="return-header"
              role="button"
              tabindex="0"
            >
              Returns
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              Returns
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              Returns
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1orzuox-MuiTableCell-root"
            scope="col"
          >
            <span
              class="MuiButtonBase-root MuiTableSortLabel-root MuiTableSortLabel-directionAsc css-7x9vt0-MuiButtonBase-root-MuiTableSortLabel-root"
              role="button"
              tabindex="0"
            >
              Returns
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-7ojauk-MuiSvgIcon-root-MuiTableSortLabel-icon"
                data-testid="ArrowDownwardIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                />
              </svg>
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          data-testid="Group A-expand"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
            </button>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            <strong>
              Group A
            </strong>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            10
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            2
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            2
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            colspan="7"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
          data-testid="Group B-expand"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-aixiz9-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
                />
              </svg>
            </button>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            <strong>
              Group B
            </strong>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            20
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            4
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
          >
            4
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1dc80h3-MuiTableCell-root"
            colspan="7"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
      <tfoot
        class="MuiTableFooter-root css-1mgfs7f-MuiTableFooter-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-footer css-1xtozoh-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-134h4iv-MuiTableCell-root"
            colspan="2"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            >
              Totals
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-134h4iv-MuiTableCell-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            >
              30
            </div>
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium css-134h4iv-MuiTableCell-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1wmifpn-MuiTypography-root"
            >
              6
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
